lib/src/diff.rs

```diff
@@ -313,23 +313,28 @@ fn unchanged_ranges_lcs(
     result
 }
 
+// Structurally the same as Range, but not implement invalid ops like iter().
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct RangeOffset<T> {
+    pub start: T,
+    pub end: T,
+}
+
 #[derive(Clone, PartialEq, Eq, Debug)]
 struct UnchangedRange {
     base_range: Range<usize>,
-    offsets: Vec<isize>,
+    offsets: Vec<RangeOffset<isize>>,
```

Comment:
Now that the base ranges and the other ranges are more independent because we don't enforce that their lengths match, is it simpler to have a single `ranges: Vec<Range<usize>>`?

Reply:
Good point, and it revealed a bigger problem that `intersect_regions()` can split existing other ranges at non-token boundary. Perhaps, it has to be based on token indices, but that might be more complicated than preprocessing inputs.

Reply:
> it has to be based on token indices, but that might be more complicated than preprocessing inputs.

Actually, it is simpler because intersection of ranges will be a basic merge join of indices. I'll send new PR separately.

Extracted some patches to https://github.com/martinvonz/jj/pull/4527. This PR also contains non-controversial changes to prepare for index-based range intersection.

Reply:
Updated to use absolute other ranges. `compact_unchanged_ranges()` should also be fixed.